### PR TITLE
Port ppx from andreypopp/ppx_deriving

### PR DESCRIPTION
**Pull request by jchavarri**
This PR ports @andreypopp's [PPX](https://github.com/andreypopp/ppx_deriving/tree/2b6707edceef9f637c342521009f4be00886d001/json) to this repo.

~For now, the current state is "`dune build` works" 😅 . But it's a good point to open discussion.~ Edit: all issues have now been fixed.

### Proposal

The idea is to offer the json PPX as part of the `melange-json` package. As an addition, native compatibility is offered as well through Yojson.

Having everything in a single repo / package makes maintenance easier for maintainers, and usage more straight forward for users.

(As a nice side effect, we got GH actions workflow for opam 💰 )

### Questions

- Should we modify the native ppx so that it works with a flag `-native`? I personally think it's cleaner to leave the current approach with its own ppx / exe
- Are internal names for libraries ok? e.g. `ppx_deriving_json_browser`
- Are `browser` and `native` suffixes (suffices?) ok?

### TODO

- [x] Add changelog entry
- [x] Fix nix build
- [x] Fix 4.14 build

### CODE REVIEW

File: .ocamlformat
Comment (anmonteiro) [inline .ocamlformat:2]: why 74 columns specifically?

File: ppx/browser/dune
Comment (anmonteiro) [inline ppx/browser/dune:2]: should we name this `melange-json.js` instead?

File: ppx/browser/dune
Comment (anmonteiro) [inline ppx/browser/dune:19]: is this a wrapped library to reuse the `ppx_deriving_json_runtime` module?

File: ppx/browser/dune
Comment (anmonteiro) [inline ppx/browser/dune:22]: could we use the `test` stanza here?

File: ppx/browser/dune
Comment (anmonteiro) [inline ppx/browser/dune:9]: could we remove the `containers` dependency? What's the biggest argument in favor of keeping it?

Comment (anmonteiro) [reply ppx/browser/dune]: I did this in https://github.com/melange-community/melange-json/pull/10/commits/b5aa8a5cab2c6de96e36f7241f1b6c856929e114

**Comment by anmonteiro** [comment]
> Should we modify the native ppx so that it works with a flag -native? I personally think it's cleaner to leave the current approach with its own ppx / exe

I think it's alright to leave it as is

> Are internal names for libraries ok? e.g. ppx_deriving_json_browser

I commented this elsewhere but I think `browser` doesn't always convey "all the places JS runs", but rather only a subset.

> Are browser and native suffixes (suffices?) ok?

I think I prefer `js`.

File: ppx/browser/dune
Comment (andreypopp) [reply ppx/browser/dune:19]: yes, so the same code could be compiled for native and js

File: .ocamlformat
Comment (andreypopp) [reply .ocamlformat:2]: my preference, feel free to reformat

File: ppx/browser/dune
Comment (andreypopp) [reply ppx/browser/dune:22]: not really, this thing makes a standalone driver, was useful for quick ad-hoc tests but feel free to remove it

**Comment by andreypopp** [comment]
Another thing might be worth looking at (but can do later) is to trim `Ppx_deriving_tools` module, it has more code than needed for this ppx and more things exposed than needed for this ppx.

**Comment by jchavarri** [comment]
I renamed all appearances of `browser` to `js` in https://github.com/melange-community/melange-json/pull/10/commits/460e4a40328827e18cbdb13a2b0e6593835f25a2.

> Another thing might be worth looking at (but can do later) is to trim `Ppx_deriving_tools` module, it has more code than needed for this ppx and more things exposed than needed for this ppx.

I tried cleaning this up but didn't see any obvious / easy things to remove. For now I just moved `tools` folder inside `ppx` to keep things tidy, as this folder is exclusively needed in the ppx.

I also updated the ppx readme to remove custom opam repo addition as it won't be needed anymore, added a changelog entry, and fixed the 4.14 compatibility.

The last remaining thing that I can think of is fixing CI for nix, @anmonteiro could you give me a hand with that please? 🥺

**Comment by andreypopp** [comment]
another thought on naming, maybe for ppx it's better to have ppx in its name?
- `melange-json.ppx` for js ppx
- `melange-json.ppx-native` for native ppx
- `melange-json.ppx-runtime` for runtime (both js and native)


I quite often use `ocamlfind list | rg ...` to see what's available and having ppx in there will hint how to use the lib

**Comment by jchavarri** [comment]
Yes, I like that naming much more. It's also consistent with other ppxs in Melange ecosystem, like the pervasive `melange.ppx`.

**Comment by jchavarri** [comment]
@andreypopp I did the renaming in https://github.com/melange-community/melange-json/pull/10/commits/d3c6e5dfb074685a21479d55c0f06d9be341cd97. 

> `melange-json.ppx-runtime` for runtime (both js and native)

I wasn't sure how to use the same name for both (there can only exist one public lib name at the moment), so I went with `melange-json.ppx-runtime` for js, and `melange-json.ppx-native-runtime` for native.

**Comment by jchavarri** [comment]
> and `melange-json.ppx-native-runtime` for native.

Actually `melange-json.ppx-runtime-native` after last commit 😓

**Comment by andreypopp** [comment]
ah, yeah, not possible to have clashes in public names, but not a big deal, these libs will be added by dune automatically once you use a corresponding ppx

**Comment by anmonteiro** [comment]
I'm also wondering what your preference is regarding the ppx being its own package, e.g. 

- `melange-json` for the melange library
- `melange-json-ppx` for the ppx? and the library names could be:
    - `melange-json-ppx` for the melange ppx
    - `melange-json-ppx.native` for the native ppx

No strong preference here, just providing an alternative

**Comment by andreypopp** [comment]
few more points:
- need to unify exceptions between melange-json and the ppx, I think this is important
- let's leave the native part undocumented for now, I think we'd want to transition to Yojson.Safe.t before make it public

**Comment by andreypopp** [comment]
> I'm also wondering what your preference is regarding the ppx being its own package

other options:
- have js ppx in melange-json package but split away native ppx into a separate package
- split js ppx and native ppx into separate packages (3 packages total counting melange-json)

the nice thing about splitting away native ppx is that we won't impose yojson dep on melange users and we won't impose melange dep on native users

**Comment by jchavarri** [comment]
> need to unify exceptions between melange-json and the ppx, I think this is important

@andreypopp could you elaborate on how this unification would look like? `melange-json` has `DecodeError` and `ParseError`. the new ppx has `Error`, which is used exclusively for the "Not supported" messages. I don't see how they could be unified?

> > I'm also wondering what your preference is regarding the ppx being its own package
> 
> other options:
> 
>     * have js ppx in melange-json package but split away native ppx into a separate package
> 
>     * split js ppx and native ppx into separate packages (3 packages total counting melange-json)

I prefer the "single package, until proved otherwise" approach. Having a single package makes things easier for maintainers and users. The "until proved otherwise" means that if some user some day (or ourselves), comes with a legit use case where separate packages are needed, then we can always split. Otherwise, it seems like a preoptimization.

A couple of examples where multi package made things harder for us in the past (in this case, it's `reason-react` and `reason-react-ppx`):
- As maintainers, publishing in opam might lead to cycles in dependencies between the packages, [example](https://github.com/ocaml/opam-repository/pull/24422#discussion_r1325419308)
- As users, I spent (just yesterday) 20min blocked on some issue with the code not updating, to later realize I was pinning `reason-react`, but not `reason-react-ppx`

Maybe I am missing what are the main upsides of having multiple packages? The "worst" part of single package is that as a user I am downloading and installing deps that I am not using, but all things considered it doesn't outweight the downsides imo.

**Comment by jchavarri** [comment]
> let's leave the native part undocumented for now, I think we'd want to transition to Yojson.Safe.t before making it public

Done in https://github.com/melange-community/melange-json/pull/10/commits/7b7e79250286f0f8221d77f4e3a405863ee1b85b.

**Comment by andreypopp** [comment]
> @andreypopp could you elaborate on how this unification would look like? melange-json has DecodeError and ParseError. the new ppx has Error, which is used exclusively for the "Not supported" messages. I don't see how they could be unified?

@jchavarri the ppx runtime uses [this](https://github.com/melange-community/melange-json/pull/10/files#diff-3c200b422db8e196e232ba368669730afb0f9d3388c2677d203611626a192224R7) exception and melange-json has [DecodeError](https://github.com/melange-community/melange-json/blob/main/src/Json_decode.mli#L16), so I think ppx runtime should also use DecodeError — that way if users would want to catch the exception and recover — they will only have to do this for DecodeError.

**Comment by andreypopp** [comment]
>>let's leave the native part undocumented for now, I think we'd want to transition to Yojson.Safe.t before making it public

> Done in https://github.com/melange-community/melange-json/commit/3a31039c7c46c5d0f0b6a44da180fda96275be30.

The README still mentions yojson/native. By the way, should we move docs from this README to the main project README, why have two?

**Comment by jchavarri** [comment]
> the ppx runtime uses [this](https://github.com/melange-community/melange-json/pull/10/files#diff-3c200b422db8e196e232ba368669730afb0f9d3388c2677d203611626a192224R7) exception and melange-json has [DecodeError](https://github.com/melange-community/melange-json/blob/main/src/Json_decode.mli#L16), so I think ppx runtime should also use DecodeError — that way if users would want to catch the exception and recover — they will only have to do this for DecodeError.

updated the ppx to use `DecodeError` in https://github.com/melange-community/melange-json/pull/10/commits/b02fb3c3bb8ddaaa60054bc59ed51e0fb27e5fee. pls lmk if this is what you had in mind.

> The README still mentions yojson/native. By the way, should we move docs from this README to the main project README, why have two?

both should be fixed in https://github.com/melange-community/melange-json/pull/10/commits/b02fb3c3bb8ddaaa60054bc59ed51e0fb27e5fee.

**Review by andreypopp — APPROVED** [review]
All looks good to me, thank you!

**Pull request merged**
